libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,674 @@
+
+   * IBusBM
+     https://github.com/bmellink/IBusBM
+
+   * BetaFlight
+     https://github.com/betaflight/betaflight/blob/master/src/main/telemetry/ibus_shared.c
+ */
+
+#include <AP_IBus_Telem/AP_IBus_Telem.h>
+
+#if AP_IBUS_TELEM_ENABLED
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Vehicle/AP_Vehicle.h>
+#include <AP_BattMonitor/AP_BattMonitor.h>
+#include <AP_Arming/AP_Arming.h>
+#include <AP_GPS/AP_GPS.h>
+#include <AP_RPM/AP_RPM.h>
+
+// All the sensors we can accurately provide are listed here.
+// i-BUS will generally only query up to 15 sensors, so subjectively
+// higher-value sensors are sorted to the top to make the most of a
+// small telemetry window. In the future these could be configurable.
+static const AP_IBus_Telem::SensorDefinition _sensors[] = {
+    {.sensor_type = IBUS_SENSOR_TYPE_ARMED, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_FLIGHT_MODE, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_GPS_STATUS, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_FUEL, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_EXTERNAL_VOLTAGE, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_ALT, .sensor_length = 4},
+    {.sensor_type = IBUS_SENSOR_TYPE_GPS_DIST, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_CLIMB_RATE, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_GROUND_SPEED, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_ROLL, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_PITCH, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_YAW, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_SPEED, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_TEMPERATURE_PRESSURE, .sensor_length = 4},
+#if AP_RPM_ENABLED == 1
+    {.sensor_type = IBUS_SENSOR_TYPE_RPM, .sensor_length = 2},
+#endif
+    {.sensor_type = IBUS_SENSOR_TYPE_BATTERY_CURRENT, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_AVERAGE_CELL_VOLTAGE, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_COMPASS_HEADING, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_COG, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_GPS_LAT, .sensor_length = 4},
+    {.sensor_type = IBUS_SENSOR_TYPE_GPS_LNG, .sensor_length = 4},
+    {.sensor_type = IBUS_SENSOR_TYPE_GPS_ALT, .sensor_length = 4},
+    {.sensor_type = IBUS_SENSOR_TYPE_ACC_X, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_ACC_Y, .sensor_length = 2},
+    {.sensor_type = IBUS_SENSOR_TYPE_ACC_Z, .sensor_length = 2},
+};
+
+/* Ground vehicle modes:
+    MANUAL       = 0
+    ACRO         = 1
+    STEERING     = 3
+    HOLD         = 4
+    LOITER       = 5
+    FOLLOW       = 6
+    SIMPLE       = 7
+    DOCK         = 8
+    CIRCLE       = 9
+    AUTO         = 10
+    RTL          = 11
+    SMART_RTL    = 12
+    GUIDED       = 15
+    INITIALISING = 16
+*/
+static const AP_IBus_Telem::ModeMap _ground_vehicle_mode_map[] = {
```

Comment:
```suggestion
#if APM_BUILD_TYPE(APM_BUILD_ArduPlane)
static const AP_IBus_Telem::ModeMap mode_map[] = {
.
.
};
#elif APM_BUILD_TYPE(APM_BUILD_ArduCopter)
static const AP_IBus_Telem::ModeMap mode_map[] = {
.
.
```

That way only ever have one `mode_map` to look through (avoiding the switch on the AHRS concept of the vehicle)

Reply:
I did originally build it this way but I was getting `@Invalid_use_of_APM_BUILD_TYPE` errors. Maybe that'll have changed now it's in the .cpp file? I'll give it a go.

Reply:
Yep - the defines I mentioned aren't allowed to be used in headers which aren't on a difficult-to-find whitelist.  It's done this way to try to keep the compilation the same for different vehicles, leading to faster build-times.